File: PracticeAssignment2Part1/Q5_Stack.java
package PracticeAssignment2Part1;

public class Q5_Stack {
    class Node{
        int data;
        Node next;
        Node prev;
        public Node(int data){
            this.data = data;
            next = null;
            prev = null;
        }
    }
    Node topStack;
    public int top(){
        if(topStack==null){
            return -1;// throwing exception

        }
        return topStack.data;
    }
    public void push(int x){
        Node newNode = new Node(x);
        if(topStack==null){
            topStack = newNode;
        }else{
            newNode.prev = topStack;
            topStack.next = newNode;
            topStack = newNode;
        }
        return;
    }
    public int pop(){
        if(topStack==null){
            return -1; // thorwing exception
        }
        Node newNode = topStack;
        topStack = topStack.prev;
        return newNode.data;
        
    }
    public boolean isEmpty(){
        return topStack == null;
    }


}

File: homework2/jeffery_owusu_agyemang/q5_Stack.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Stack:
    def __init__(self):
        self.top_node = None  # top of stack

    # O(1)
    def top(self):
        if not self.top_node:
            return None
        return self.top_node.data

    # O(1)
    def push(self, x):
        new_node = Node(x)
        new_node.next = self.top_node
        self.top_node = new_node

    # O(1)
    def pop(self):
        if not self.top_node:
            return None
        val = self.top_node.data
        self.top_node = self.top_node.next
        return val

    # O(1)
    def isEmpty(self):
        return self.top_node is None